File: android/core/src/main/kotlin/com/dienbell/tactics/core/chess/Board.kt
package com.dienbell.tactics.core.chess

import kotlin.math.abs

/**
 * A chess position: piece placement plus the side to move. FEN castling/en
 * passant/halfmove counters are intentionally not modelled — only piece
 * placement and side to move are parsed, matching the iOS engine.
 */
class Board private constructor(
    initialPieces: Map<Square, Piece>,
    val sideToMove: PieceColor,
) {
    var pieces: Map<Square, Piece> = initialPieces
        private set

    sealed class FENError(message: String) : Exception(message) {
        data object MissingFields : FENError("FEN is missing fields")
        data object InvalidRankCount : FENError("FEN has an invalid rank count")
        class InvalidRank(rank: String) : FENError("FEN has an invalid rank: $rank")
        class InvalidPiece(piece: Char) : FENError("FEN has an invalid piece: $piece")
        class InvalidSideToMove(side: String) : FENError("FEN has an invalid side to move: $side")
    }

    /** Applies `move` (with castling, en passant, promotion). Returns false if no
     *  piece sits on the origin square. */
    fun apply(move: ChessMove): Boolean {
        val moving = pieces[move.from] ?: return false
        val board = pieces.toMutableMap()
        board.remove(move.from)
        val targetWasEmpty = board[move.to] == null
        val resultingKind = move.promotion ?: moving.kind
        board[move.to] = Piece(moving.color, resultingKind)

        // Castling: a king advancing two files also relocates the corner rook.
        if (moving.kind == PieceKind.KING && abs(move.to.file - move.from.file) == 2) {
            val rank = move.from.rank
            when (move.to.file) {
                6 -> board.remove(Square(7, rank))?.let { rook -> board[Square(5, rank)] = rook }
                2 -> board.remove(Square(0, rank))?.let { rook -> board[Square(3, rank)] = rook }
            }
        }

        // En passant: a pawn moving diagonally onto an empty square captures the
        // enemy pawn that just advanced beside its origin.
        if (moving.kind == PieceKind.PAWN && move.from.file != move.to.file && targetWasEmpty) {
            board.remove(Square(move.to.file, move.from.rank))
        }

        pieces = board
        return true
    }

    /** Basic piece-movement legality for `mover`: shape, blocking, self-capture.
     *  Does not cover check/pin, castling, en passant, or promotion. */
    fun isLegal(move: ChessMove, mover: PieceColor): Boolean {
        val piece = pieces[move.from] ?: return false
        if (piece.color != mover) return false
        val occupant = pieces[move.to]
        if (occupant != null && occupant.color == mover) return false
        return shapeIsLegal(piece, move.from, move.to)
    }

    private fun shapeIsLegal(piece: Piece, from: Square, to: Square): Boolean {
        val df = to.file - from.file
        val dr = to.rank - from.rank
        if (df == 0 && dr == 0) return false
        return when (piece.kind) {
            PieceKind.KNIGHT ->
                (abs(df) == 1 && abs(dr) == 2) || (abs(df) == 2 && abs(dr) == 1)
            PieceKind.BISHOP ->
                abs(df) == abs(dr) && pathIsClear(from, to)
            PieceKind.ROOK ->
                (df == 0 || dr == 0) && pathIsClear(from, to)
            PieceKind.QUEEN ->
                (df == 0 || dr == 0 || abs(df) == abs(dr)) && pathIsClear(from, to)
            PieceKind.KING ->
                maxOf(abs(df), abs(dr)) == 1
            PieceKind.PAWN ->
                pawnShapeIsLegal(piece, from, to)
        }
    }

    private fun pawnShapeIsLegal(pawn: Piece, from: Square, to: Square): Boolean {
        val df = to.file - from.file
        val dr = to.rank - from.rank
        val direction = if (pawn.color == PieceColor.WHITE) 1 else -1
        val startRank = if (pawn.color == PieceColor.WHITE) 1 else 6

        if (df == 0) {
            if (dr == direction && pieces[to] == null) return true
            if (from.rank == startRank && dr == 2 * direction) {
                val middle = Square(from.file, from.rank + direction)
                return pieces[to] == null && pieces[middle] == null
            }
            return false
        }
        if (abs(df) == 1 && dr == direction) {
            return pieces[to]?.color == pawn.color.opponent
        }
        return false
    }

    private fun pathIsClear(from: Square, to: Square): Boolean {
        val sf = signum(to.file - from.file)
        val sr = signum(to.rank - from.rank)
        var file = from.file + sf
        var rank = from.rank + sr
        while (file != to.file || rank != to.rank) {
            if (pieces[Square(file, rank)] != null) return false
            file += sf
            rank += sr
        }
        return true
    }

    private fun signum(x: Int): Int = when {
        x > 0 -> 1
        x < 0 -> -1
        else -> 0
    }

    companion object {
        fun parse(fen: String): Board {
            val fields = fen.split(" ")
            if (fields.size < 2) throw FENError.MissingFields

            val ranks = fields[0].split("/")
            if (ranks.size != 8) throw FENError.InvalidRankCount

            val pieces = mutableMapOf<Square, Piece>()
            for ((rankIndex, encodedRank) in ranks.withIndex()) {
                var file = 0
                for (ch in encodedRank) {
                    val empty = ch.digitToIntOrNull()
                    if (empty != null) {
                        if (empty !in 1..8) throw FENError.InvalidRank(encodedRank)
                        file += empty
                    } else {
                        if (file >= 8) throw FENError.InvalidPiece(ch)
                        val piece = pieceFor(ch) ?: throw FENError.InvalidPiece(ch)
                        pieces[Square(file, 7 - rankIndex)] = piece
                        file += 1
                    }
                }
                if (file != 8) throw FENError.InvalidRank(encodedRank)
            }

            val sideToMove = when (fields[1]) {
                "w" -> PieceColor.WHITE
                "b" -> PieceColor.BLACK
                else -> throw FENError.InvalidSideToMove(fields[1])
            }
            return Board(pieces, sideToMove)
        }

        private fun pieceFor(ch: Char): Piece? {
            val color = if (ch.isUpperCase()) PieceColor.WHITE else PieceColor.BLACK
            val kind = when (ch.lowercaseChar()) {
                'k' -> PieceKind.KING
                'q' -> PieceKind.QUEEN
                'r' -> PieceKind.ROOK
                'b' -> PieceKind.BISHOP
                'n' -> PieceKind.KNIGHT
                'p' -> PieceKind.PAWN
                else -> return null
            }
            return Piece(color, kind)
        }
    }
}

File: ios/ChessCore/Sources/Board.swift
import Foundation

public struct Board: Equatable, Sendable {
    public enum FENError: Error, Equatable {
        case missingFields
        case invalidRankCount
        case invalidRank(String)
        case invalidPiece(Character)
        case invalidSideToMove(String)
    }

    /// Castling availability parsed from the FEN's third field ("KQkq" style).
    /// Empty when the FEN omits or blanks the field.
    public struct CastlingRights: OptionSet, Equatable, Sendable {
        public let rawValue: Int
        public init(rawValue: Int) { self.rawValue = rawValue }
        public static let whiteKingSide = CastlingRights(rawValue: 1 << 0)
        public static let whiteQueenSide = CastlingRights(rawValue: 1 << 1)
        public static let blackKingSide = CastlingRights(rawValue: 1 << 2)
        public static let blackQueenSide = CastlingRights(rawValue: 1 << 3)
    }

    public private(set) var pieces: [Square: Piece]
    public private(set) var sideToMove: PieceColor
    /// FEN fields 3–4. Kept for validation and special-move handling; the
    /// halfmove clock and fullmove number (fields 5–6) are not needed for
    /// tactics lines and are not stored.
    public private(set) var castlingRights: CastlingRights
    /// The square where en passant capture is possible, if any.
    public private(set) var enPassantTarget: Square?
    /// Set by `apply` when a move leaves the mover's king attacked. The board
    /// is a passive record of the position; check detection lives here so
    /// `isLegal` can reject self-checks and `apply` stays unchecked.
    public private(set) var sideToMoveInCheck: Bool = false

    public init(fen: String) throws {
        let fields = fen.split(separator: " ")
        guard fields.count >= 2 else { throw FENError.missingFields }

        let ranks = fields[0].split(separator: "/", omittingEmptySubsequences: false)
        guard ranks.count == 8 else { throw FENError.invalidRankCount }

        var parsedPieces: [Square: Piece] = [:]
        for (rankIndex, encodedRank) in ranks.enumerated() {
            var file = 0
            var previousWasDigit = false
            for character in encodedRank {
                // Only a single ASCII digit 1–8 is a rank skip. Non-ASCII
                // digits ("٣") and multi-digit runs ("44") are rejected rather
                // than summed or silently accepted.
                if let emptyCount = character.wholeNumberValue, character.isASCII {
                    guard (1...8).contains(emptyCount), !previousWasDigit else {
                        throw FENError.invalidRank(String(encodedRank))
                    }
                    file += emptyCount
                    previousWasDigit = true
                } else {
                    guard file < 8, let piece = Self.piece(for: character) else {
                        throw FENError.invalidPiece(character)
                    }
                    parsedPieces[Square(file: file, rank: 7 - rankIndex)!] = piece
                    file += 1
                    previousWasDigit = false
                }
            }
            guard file == 8 else { throw FENError.invalidRank(String(encodedRank)) }
        }

        switch fields[1] {
        case "w": sideToMove = .white
        case "b": sideToMove = .black
        default: throw FENError.invalidSideToMove(String(fields[1]))
        }
        castlingRights = Self.castlingRights(fields.count > 2 ? String(fields[2]) : "")
        enPassantTarget = fields.count > 3 ? Square(notation: String(fields[3])) : nil
        pieces = parsedPieces
        sideToMoveInCheck = Self.isAttacked(
            kingSquare(of: sideToMove, in: parsedPieces),
            by: sideToMove.opponent,
            in: parsedPieces,
            enPassantTarget: nil
        )
    }

    private static func castlingRights(_ field: String) -> CastlingRights {
        var rights: CastlingRights = []
        if field.contains("K") { rights.insert(.whiteKingSide) }
        if field.contains("Q") { rights.insert(.whiteQueenSide) }
        if field.contains("k") { rights.insert(.blackKingSide) }
        if field.contains("q") { rights.insert(.blackQueenSide) }
        return rights
    }

    public mutating func apply(_ move: ChessMove) -> Bool {
        guard let movingPiece = pieces.removeValue(forKey: move.from) else { return false }
        let targetWasEmpty = pieces[move.to] == nil
        let resultingKind = move.promotion ?? movingPiece.kind
        pieces[move.to] = Piece(color: movingPiece.color, kind: resultingKind)

        // Castling: a king advancing two files also relocates the corner rook.
        // The rook must actually be there; otherwise the king simply moved
        // (a data bug, not a position we should corrupt with a phantom rook).
        if movingPiece.kind == .king, abs(move.to.file - move.from.file) == 2 {
            let rank = move.from.rank
            if move.to.file == 6, let rook = pieces.removeValue(forKey: Square(file: 7, rank: rank)!) {
                pieces[Square(file: 5, rank: rank)!] = rook
            } else if move.to.file == 2, let rook = pieces.removeValue(forKey: Square(file: 0, rank: rank)!) {
                pieces[Square(file: 3, rank: rank)!] = rook
            }
        }

        // En passant: a pawn moving diagonally onto an empty square captures
        // the enemy pawn beside its origin. Only the square an en-passant
        // capture can actually remove — an enemy pawn on it — is cleared.
        if movingPiece.kind == .pawn,
           move.from.file != move.to.file,
           targetWasEmpty {
            let capturedSquare = Square(file: move.to.file, rank: move.from.rank)!
            if pieces[capturedSquare]?.color == movingPiece.color.opponent {
                pieces.removeValue(forKey: capturedSquare)
            }
        }

        // Toggle the recorded turn and update the special-move context so the
        // next `isLegal` call answers for the side about to move.
        sideToMove = sideToMove.opponent
        enPassantTarget = (movingPiece.kind == .pawn && abs(move.to.rank - move.from.rank) == 2)
            ? Square(file: move.from.file, rank: (move.from.rank + move.to.rank) / 2)!
            : nil
        if !castlingRights.isEmpty {
            if movingPiece.kind == .king {
                let whiteRights: CastlingRights = [.whiteKingSide, .whiteQueenSide]
                let blackRights: CastlingRights = [.blackKingSide, .blackQueenSide]
                castlingRights.subtract(movingPiece.color == .white ? whiteRights : blackRights)
            }
            if movingPiece.kind == .rook {
                let backRank = movingPiece.color == .white ? 0 : 7
                if move.from == Square(file: 0, rank: backRank)! {
                    castlingRights.subtract(movingPiece.color == .white ? .whiteQueenSide : .blackQueenSide)
                } else if move.from == Square(file: 7, rank: backRank)! {
                    castlingRights.subtract(movingPiece.color == .white ? .whiteKingSide : .blackKingSide)
                }
            }
        }
        sideToMoveInCheck = Self.isAttacked(
            kingSquare(of: sideToMove, in: pieces),
            by: sideToMove.opponent,
            in: pieces,
            enPassantTarget: enPassantTarget
        )
        return true
    }

    /// `apply` minus the bookkeeping `isLegal` doesn't need. Kept separate so
    /// the legality probe cannot observe intermediate rights/turn updates.
    mutating func applyUncheckedForLegality(_ move: ChessMove) -> Bool {
        guard let movingPiece = pieces.removeValue(forKey: move.from) else { return false }
        let targetWasEmpty = pieces[move.to] == nil
        pieces[move.to] = Piece(color: movingPiece.color, kind: move.promotion ?? movingPiece.kind)

        if movingPiece.kind == .king, abs(move.to.file - move.from.file) == 2 {
            let rank = move.from.rank
            if move.to.file == 6, let rook = pieces.removeValue(forKey: Square(file: 7, rank: rank)!) {
                pieces[Square(file: 5, rank: rank)!] = rook
            } else if move.to.file == 2, let rook = pieces.removeValue(forKey: Square(file: 0, rank: rank)!) {
                pieces[Square(file: 3, rank: rank)!] = rook
            }
        }

        if movingPiece.kind == .pawn, move.from.file != move.to.file, targetWasEmpty {
            let capturedSquare = Square(file: move.to.file, rank: move.from.rank)!
            if pieces[capturedSquare]?.color == movingPiece.color.opponent {
                pieces.removeValue(forKey: capturedSquare)
            }
        }
        return true
    }


    func kingSquare(of color: PieceColor, in pieces: [Square: Piece]) -> Square? {
        pieces.first { _, piece in piece.color == color && piece.kind == .king }?.key
    }

    /// Whether any piece of `attacker`'s color attacks `square` in the classic
    /// sense (capture moves, ignoring en-passant-only oddities).
    static func isAttacked(
        _ square: Square?,
        by attacker: PieceColor,
        in pieces: [Square: Piece],
        enPassantTarget: Square?
    ) -> Bool {
        guard let square else { return false }

        for (origin, piece) in pieces where piece.color == attacker {
            switch piece.kind {
            case .pawn:
                let direction = attacker == .white ? 1 : -1
                if abs(square.file - origin.file) == 1, square.rank - origin.rank == direction { return true }
            case .knight:
                if (abs(square.file - origin.file), abs(square.rank - origin.rank)) == (1, 2)
                    || (abs(square.file - origin.file), abs(square.rank - origin.rank)) == (2, 1) { return true }
            case .king:
                if max(abs(square.file - origin.file), abs(square.rank - origin.rank)) == 1 { return true }
            case .bishop, .rook, .queen:
                if shapeIsLegal(piece, from: origin, to: square, in: pieces) {
                    // A king adjacent to the checked king still attacks it even
                    // though the reverse would be illegal; shapeIsLegal already
                    // covers that, so the shape test alone is sufficient here.
                    return true
                }
            }
        }
        return false
    }

    private static func piece(for character: Character) -> Piece? {
        let color: PieceColor = character.isUppercase ? .white : .black
        let kind: PieceKind
        switch character.lowercased() {
        case "k": kind = .king
        case "q": kind = .queen
        case "r": kind = .rook
        case "b": kind = .bishop
        case "n": kind = .knight
        case "p": kind = .pawn
        default: return nil
        }
        return Piece(color: color, kind: kind)
    }
}
